README.md

```diff
@@ -142,16 +142,16 @@ dependencies {
 ```groovy
 // Add Dagger dependencies
 dependencies {
-  compile 'com.google.dagger:dagger:2.x'
+  implementation 'com.google.dagger:dagger:2.x'
```

Comment:
For libraries you'd most likely use `api` right? (because you'll expose Dagger modules, or classes that use `@Inject` –possibly even Dagger's `Lazy`– that you'll then need Dagger to see and process in downstream projects using Dagger components)
And for applications, `api` vs. `implementation` shouldn't really make a difference.

So I'd rather switch to `api` (assuming the goal here is to silence the deprecation warning when using `compile`)

_Disclosure: I'm not an Android developer._

Reply:
I'd say it depends on how you build your apps.

I have some where the manually written component is the only public API and the modules and instantiation is entirely encapsulated. The annotations such as `@Component` and `@BindsInstance` are on that API, but they're not relevant to consumers of the component and its builder.

But on the other hand I've had ones where types like `Provider` and `Lazy` were exposed for constructor injection by downstream modules to hook up.

While the affordances provided by `api` vs. `implementation` are a pleasure to have as a Gradle user, as a library developer it's nothing short of a giant pain in the ass. Every readme is now required to have a mini tutorial in explaining module encapsulation. And usually it's far too nuanced to explain properly to people who just want to copy/paste their way to something to play around with.

Reply:
> While the affordances provided by api vs. implementation are a pleasure to have as a Gradle user, as a library developer it's nothing short of a giant pain in the ass. Every readme is now required to have a mini tutorial in explaining module encapsulation. And usually it's far too nuanced to explain properly to people who just want to copy/paste their way to something to play around with.

If you see `implementation` as mostly a build-time optimization (which it really is: classpath for compiling downstream modules is smaller, dependencies aren't checked for API changes and won't trigger a rebuild of downstream modules), then the copy/paste version should be `api`; and let people who care about build time and/or encapsulation investigate and switch to `implementation` if that works for them.

See also recent discussions in AutoValue, and older ones in Guava, about exposing annotations downstream when they're applied on API elements.

Reply:
My motivation for opening the PR: I was starting a new sample project.  As I was adding my common dependencies I noticed the readmes seem to lean towards either still having `compile` or going towards `implementation`:
https://github.com/ReactiveX/RxAndroid
https://github.com/square/moshi
https://github.com/JakeWharton/timber

Reply:
I don't disagree, although I'll also wager almost everyone who copy/paste these things don't have modularized applications in the first place.

The problem is the tooling tells you to use 'implementation' to replace 'compile' despite not being equivalent.

Reply:
It's also worth mentioning that compile and api are technically the same at the moment.

Reply:
Sorry, I thought I had commented here, but clearly I hadn't...

I agree with a lot of what is said here. In general, I don't like how every project in the world specifies the ways that people can depend on their project with maven, gradle, gradle-android, bazel, buck, etc. While it's currently been a simple reprieve to all of the questions, it's a road down a bad path as we're seeing now. It also is unnecessary boilerplate for every single library in the open.

Ideally, projects would declare their maven coordinates and users will spend a few minutes of their time to understand a little about the particular build system that they use. But we do need something in the interim.

I think the only safe thing to do here is to declare these as `api` and trust that people who care about optimizing their build time will read more about `api` vs. `implementation` and make a selection based on their findings. We can help point people in that way too.

If we were to do `implementation` as the default recommendation, I can already imagine the swaths of questions that will start flowing in asking why dagger is broken on Android 22.5 and Samsung.

@digitalbuddha all that being said - what about switching to `api` and adding a `gradle.md` file that we link to and describes this at a little greater depth?

Reply:
Sounds good. Give me a few days to update. Mid vacation 😁

Any thoughts on what the in depth discussion would be? Is there an android resource we can link to?

Reply:
Here is the output with the link that the official tools spit out when `compile` is used. 

> WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
> It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html